src/libsync/propagatecommonzsync.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) by Ahmed Ammar <ahmed.a.ammar@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include <QLoggingCategory>
+#include <QTemporaryFile>
+
+extern "C" {
+    #include "librcksum/rcksum.h"
+    #include "libzsync/zmap.h"
+    #include "libzsync/sha1.h"
+    #include "libzsync/zsync.h"
+    #include "format_string.h"
+}
+
+namespace OCC {
+Q_DECLARE_LOGGING_CATEGORY(lcZsyncPut)
+Q_DECLARE_LOGGING_CATEGORY(lcZsyncGet)
+
+enum class ZsyncMode { download, upload };
+
+/**
+ * @ingroup libsync
+ *
+ * Function to know if we are allowed to attempt using zsync from configuration/command-line options.
+ *
+ */
+bool isZsyncPropagationEnabled(OwncloudPropagator *propagator, SyncFileItemPtr item);
+
+/**
+ * @ingroup libsync
+ *
+ * Propagation zsync seeding heleper job
+ *
+ */
+class ZsyncSeedJob : public QObject
+{
+    Q_OBJECT
+    QByteArray _zsyncData;
+    QString _zsyncFilePath;
+    ZsyncMode _type;
+public:
+    explicit ZsyncSeedJob(QByteArray &zsyncData, QString path, ZsyncMode type)
+        : _zsyncData(zsyncData)
+        , _zsyncFilePath(path)
+        , _type(type)
+    {
+    };
+
+    Q_INVOKABLE void seed();
+
+signals:
+    void finishedSignal(void *zs);
```

Comment:
Could this be a `zsync_state*`? Also add documentation saying that ownership transfers to the recipient.

There's the strong urge to wrap this pointer into something less error-prone (see missing zsync_end calls above).

Reply:
Any hints on what I can wrap it in or how that would be great.

Reply:
A `std::unique_ptr` with a custom deleter matches the desired semantics best. Qt also supports custom deleters on `QSharedPointer` -- but you don't actually need shared ownership of this resource.

Reply:
ok will look into using a `std:: unique_ptr` then.

Reply:
Done.